igor-web/src/main/groovy/com/netflix/spinnaker/igor/wercker/WerckerBuildMonitor.groovy

```diff
@@ -0,0 +1,335 @@
+        this.cache = cache
+        this.buildMasters = buildMasters
+        this.pollingEnabled = pollingEnabled
+        this.echoService = echoService
+        this.front50 = front50Service
+        this.werckerProperties = werckerProperties
+    }
+
+    @Override
+    String getName() {
+        "WerckerBuildMonitor"
+    }
+
+    @Override
+    boolean isInService() {
+        pollingEnabled && super.isInService()
+    }
+
+    @Override
+    void initialize() {
+    }
+
+    @Override
+    void poll(boolean sendEvents) {
+        long startTime = System.currentTimeMillis()
+        log.info "WerckerBuildMonitor Polling cycle started: ${new Date()}, front50:${front50.isPresent()}, echoService:${echoService.isPresent()} "
+        buildMasters.filteredMap(BuildServiceProvider.WERCKER).keySet().parallelStream().forEach( { master ->
+            pollSingle(new PollContext(master, !sendEvents))
+        }
+        )
+        log.info "WerckerBuildMonitor Polling cycle done in ${System.currentTimeMillis() - startTime}ms"
+    }
+
+    @PreDestroy
+    void stop() {
+        log.info('Stopped')
+        if (!worker.isUnsubscribed()) {
+            worker.unsubscribe()
+        }
+    }
+
+    /**
+     * Gets a list of pipelines for this master & processes runs between last poll stamp and a sliding upper bound stamp,
+     * the cursor will be used to advanced to the upper bound when all builds are completed in the commit phase.
+     */
+    @Override
+    protected PipelinePollingDelta generateDelta(PollContext ctx) {
+        String master = ctx.partitionName
+        log.info("Checking for new builds for $master")
+        def startTime = System.currentTimeMillis()
+
+        List<PipelineDelta> delta = []
+
+        WerckerService werckerService = buildMasters.map[master] as WerckerService
+        long since = System.currentTimeMillis() - (Long.valueOf(getPollInterval() * 2 * 1000))
+        try {
+            if (front50.isPresent()) {
+                Set<String> pipelines = getConfiguredJobs(master)
+                Map<String, List<Run>> runs = werckerService.getRunsSince(pipelines, since)
+                runs.keySet().forEach( { pipeline ->
+                    processRuns(werckerService, master, pipeline, delta, runs.get(pipeline))
+                } )
+            } else {
+                Map<String, List<Run>> runs = werckerService.getRunsSince(since)
+                runs.keySet().forEach( { pipeline ->
+                    processRuns(werckerService, master, pipeline, delta, runs.get(pipeline))
+                } )
+            }
+        } catch (e) {
+            log.error("Error processing runs for Wercker[{}]", kv("master", master), e)
+        }
+        log.debug("Took ${System.currentTimeMillis() - startTime}ms to retrieve Wercker pipelines (master: {})", kv("master", master))
+        return new PipelinePollingDelta(master: master, items: delta)
+    }
+
+    Set<String> getConfiguredJobs(String master) {
```

Comment:
I guess the "pre-filtering" you meant is to configure some filter in the config (igor.yml). My understanding is that the config change is done through halyard and may need to restart the entire spinnaker; restarting spinnaker would be an usability issue for production deployment.

To do the pre-filtering, we would need the list of ci triggers and stages that has been enabled in all the spinnaker pipelines. I am looking at the BuildEventMonitor but cannot find a simple way to get all ci triggers and stages. Would you clarify what in the echo that I should use and you would be more comfortable with? What I can tell is to let PipelineCache subscribes front50, polls and store the triggers and stages in the PipelineCache so that igor can query it through some new API, but that would be a new API and new logic to build, and I would consider it more complicate than my current approach. Another possible way is to get the pipelineConfigs from gate.controllers.PipelineConfigController.getAllPipelineConfigs; is this better for you?